### PR TITLE
Remove handles from ves_icall_System_Array_InternalCreate.

### PR DESCRIPTION
**Pull request by jaykrell**
Also combine two allocas.

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -818,53 +818,68 @@ ves_icall_System_Array_SetValue (MonoArrayHandle arr, MonoObjectHandle value,
...
 			lower_bounds [i] = 0;
 		}
 		sizes [i] = pLengths [i];
 	}
-	
-	return mono_array_new_full_handle (mono_domain_get (), aklass, sizes, lower_bounds, error);
+
+	*result = mono_array_new_full_checked (mono_domain_get (), aklass, sizes, lower_bounds, error);
+
+exit:
```
Comment (filipnavara) [inline mono/metadata/icall.c]: What happens if the garbage collector runs at this particular point?

Comment (vargaz) [reply mono/metadata/icall.c]: result is in the parent managed frame, so it will be scanned there.

**Comment by vargaz** [comment]
Looks ok, but its kinda ugly, is it worth it ?

**Comment by jaykrell** [comment]
It is "much" more efficient. Multiple TLS accesses are avoided -- enter/exit frame and handle create, and the "bump allocator" is replaced by address of local. Seems very worthwhile.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
I think more should be this way. Avoid the TLS-based handles in general.

File: netcore/System.Private.CoreLib/src/System/Array.Mono.cs [hunk 2]
```diff
@@ -322,16 +325,16 @@ public static unsafe Array CreateInstance (Type elementType, int[] lengths, int[
...
 			fixed (int* pLowerBounds = &lowerBounds [0])
-				array = InternalCreate (runtimeType._impl.Value, lengths.Length, pLengths, pLowerBounds);
+				InternalCreate (ref array, runtimeType._impl.Value, lengths.Length, pLengths, pLowerBounds);
 			GC.KeepAlive (runtimeType);
 			return array;
 		}
 
 		[MethodImpl (MethodImplOptions.InternalCall)]
-		static extern unsafe Array InternalCreate (IntPtr elementType, int rank, int* pLengths, int* pLowerBounds);
+		static extern unsafe void InternalCreate (ref Array result, IntPtr elementType, int rank, int* lengths, int* lowerBounds);
```
Comment (Therzok) [inline netcore/System.Private.CoreLib/src/System/Array.Mono.cs]: Shouldn't this be `out Array`?

Comment (jaykrell) [reply netcore/System.Private.CoreLib/src/System/Array.Mono.cs]: No. It should be ref.
Out is a micro optimization and
I frequently see native code fail to set out parameters, violating safety and correctness. Out is to be avoided.

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins squash

**Comment by jaykrell** [comment]
@monojenkins build failed

**Pull request merged**